synapse/logging/tracing.py

```diff
@@ -0,0 +1,944 @@
+    message: str, ret: T
+) -> Callable[[Callable[P, R]], Callable[P, Union[T, R]]]:
+    ...
+
+
+def ensure_active_span(
+    message: str, ret: Optional[T] = None
+) -> Callable[[Callable[P, R]], Callable[P, Union[Optional[T], R]]]:
+    """Executes the operation only if opentelemetry is enabled and there is an active span.
+    If there is no active span it logs message at the error level.
+
+    Args:
+        message: Message which fills in "There was no active span when trying to %s"
+            in the error log if there is no active span and opentelemetry is enabled.
+        ret: return value if opentelemetry is None or there is no active span.
+
+    Returns:
+        The result of the func, falling back to ret if opentelemetry is disabled or there
+        was no active span.
+    """
+
+    def ensure_active_span_inner_1(
+        func: Callable[P, R]
+    ) -> Callable[P, Union[Optional[T], R]]:
+        @wraps(func)
+        def ensure_active_span_inner_2(
+            *args: P.args, **kwargs: P.kwargs
+        ) -> Union[Optional[T], R]:
+            if not opentelemetry:
+                return ret
+
+            if not opentelemetry.trace.get_current_span():
+                logger.error(
+                    "There was no active span when trying to %s."
+                    " Did you forget to start one or did a context slip?",
+                    message,
+                    stack_info=True,
+                )
+
+                return ret
+
+            return func(*args, **kwargs)
+
+        return ensure_active_span_inner_2
+
+    return ensure_active_span_inner_1
+
+
+# Setup
+
+
+def init_tracer(hs: "HomeServer") -> None:
+    """Set the whitelists and initialise the OpenTelemetry tracer"""
+    global opentelemetry
+    if not hs.config.tracing.tracing_enabled:
+        # We don't have a tracer
+        opentelemetry = None  # type: ignore[assignment]
+        return
+
+    if not opentelemetry:
+        raise ConfigError(
+            "The server has been configured to use OpenTelemetry but OpenTelemetry is not "
+            "installed."
+        )
+
+    # Pull out of the config if it was given. Otherwise set it to something sensible.
+    set_homeserver_whitelist(hs.config.tracing.homeserver_whitelist)
+
+    resource = opentelemetry.sdk.resources.Resource(
+        attributes={
+            opentelemetry.sdk.resources.SERVICE_NAME: f"{hs.config.server.server_name} {hs.get_instance_name()}"
+        }
+    )
+
+    # TODO: `force_tracing_for_users` is not compatible with OTEL samplers
+    # because you can only determine `opentelemetry.trace.TraceFlags.SAMPLED`
+    # and whether it uses a recording span when the span is created and we don't
+    # have enough information at that time (we can determine in
+    # `synapse/api/auth.py`). There isn't a way to change the trace flags after
+    # the fact so there is no way to programmatically force
+    # recording/tracing/sampling like there was in opentracing.
```

Comment:
Sampling works but the only difference from this PR to to our existing functionality is that it doesn't support `force_tracing_for_users`. Because OpenTelemetry only supports "consistent upfront (or head-based) sampling" there is no way to `force_tracing` after a span has been created.

Previously, with `opentracing`, we could use the `SAMPLING_PRIORITY` tag, see https://github.com/jaegertracing/jaeger-client-python#debug-traces-forced-sampling

But with `OpenTelemetry`, the sampler decides whether to record or sample at the beginning and the tracer sets immutable [`TraceFlags`](https://opentelemetry-python.readthedocs.io/en/latest/api/trace.span.html#opentelemetry.trace.span.TraceFlags) on the span, see [`opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L1065-L1089`](https://github.com/open-telemetry/opentelemetry-python/blob/c6d7b9f3dd4864445050f86e1e5a88100f1eb84e/opentelemetry-sdk/src/opentelemetry/sdk/trace/__init__.py#L1065-L1089)

Trying to do anything different is just going against the grain. See the [spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#sampling) and doesn't seem very feasible until we get more hooks which has a proposal, see https://github.com/open-telemetry/oteps/pull/115

### Workarounds?

I attempted to workaround this by creating a custom sampler and exporter but it's [pretty jank](https://github.com/matrix-org/synapse/pull/13400/commits/b3cdbad9858aaf7155b7a3484af041863cbdc383#diff-80485970f0e6eb9a96154544e09e5e3b85af08f0a9d38ab962c8e2cf352050ddR392-R542) and doesn't solve the problem of child spans in the trace inheriting the forced tracing. It sets the `SynapseTags.FORCE_RECORD_MAYBE_SAMPLE` attribute on the top-level servlet span (any span created before we know that we should force tracing). This causes our custom sampler to return a `sampling.Decision.RECORD_ONLY` decision otherwise a `NonRecordingSpan` is used and we lose all the data before. Once we know we should force tracing based on the user, we set the `SynapseTags.FORCE_TRACING` attribute on the span/parent span and our custom exporter knows to always export those spans. The problem I am having is how to have the child spans inherit the force tracing behavior. We would ideally want them to inherit `SynapseTags.FORCE_TRACING` or set the current context to one that has sampling on since that will inherit by default.

For the inheriting, I attempted to use [baggage](https://opentelemetry-python.readthedocs.io/en/latest/api/baggage.html) (like we did with `opentracing`) for this but setting baggage on a context returns a new context instead of setting it in place and there isn't a way to change span context so it's pretty much useless AFAICT. The `opentracing` baggage is per-span and saves to the span so it sticks around and is useful to reference from other spans/contexts.

I've also thought about inheriting the `SynapseTags.FORCE_TRACING` attribute to child spans by span processor and setting it in the `on_start` span creation hook. But I have the same problem of not knowing how to determine whether to force tracing from the `parent_context` since you can't set baggage on it.

Reply:
From the [backend chapter sync last month](https://docs.google.com/document/d/1kmGRzPFfg_gRY6l0sxjYkSLW6UpMFn9ELQX5CtTLWlA/edit#bookmark=id.i7khdo7yafye), @reivilibre mentioned

> The access token contains the base64-encoded localpart of the user. That could be usable as an early indication of whether the span should be recorded or not — you’d want to extract the header before you start the top-level span. A good way to determine whether we should record before making the top-level span.

Which although is still a bit dirty, is a lot more clean than this custom sampler stuff that barely and doesn't work. A good way forward ⏩

---

It is surprising that there isn't a canonical way to do this though. There are other use cases like wanting to only record the long-running requests which you can only determine at the end.